// mongodb/src/operations/search_indexes/create.ts
import { type Document } from 'mongodb/src/bson';
import { type Connection } from 'mongodb/src/cmap/connection';
import { MongoDBResponse } from 'mongodb/src/cmap/wire_protocol/responses';
import type { Collection } from 'mongodb/src/collection';
import type { ServerCommandOptions } from 'mongodb/src/sdam/server';
import type { ClientSession } from 'mongodb/src/sessions';
import { type TimeoutContext } from 'mongodb/src/timeout';
import { AbstractOperation } from 'mongodb/src/operations/operation';

/**
 * @public
 */
export interface SearchIndexDescription extends Document {
  /** The name of the index. */
  name?: string;

  /** The index definition. */
  definition: Document;

  /** The type of the index.  Currently `search` or `vectorSearch` are supported. */
  type?: string;
}

/** @internal */
export class CreateSearchIndexesOperation extends AbstractOperation<Document> {
  override SERVER_COMMAND_RESPONSE_TYPE = MongoDBResponse;
  private readonly collection: Collection;
  private readonly descriptions: ReadonlyArray<SearchIndexDescription>;

  constructor(collection: Collection, descriptions: ReadonlyArray<SearchIndexDescription>) {
    super();
    this.collection = collection;
    this.descriptions = descriptions;
    this.ns = collection.fullNamespace;
  }

  override get commandName() {
    return 'createSearchIndexes' as const;
  }

  override buildCommand(_connection: Connection, _session?: ClientSession): Document {
    const namespace = this.collection.fullNamespace;
    return {
      createSearchIndexes: namespace.collection,
      indexes: this.descriptions
    };
  }

  override handleOk(response: InstanceType<typeof this.SERVER_COMMAND_RESPONSE_TYPE>): string[] {
    return super.handleOk(response).indexesCreated.map((val: { name: string }) => val.name);
  }

  override buildOptions(timeoutContext: TimeoutContext): ServerCommandOptions {
    return { session: this.session, timeoutContext };
  }
}
