// mongodb/src/operations/search_indexes/update.ts
import type { Document } from 'mongodb/src/bson';
import { type Connection } from 'mongodb/src/cmap/connection';
import { MongoDBResponse } from 'mongodb/src/cmap/wire_protocol/responses';
import type { Collection } from 'mongodb/src/collection';
import type { ServerCommandOptions } from 'mongodb/src/sdam/server';
import type { ClientSession } from 'mongodb/src/sessions';
import { type TimeoutContext } from 'mongodb/src/timeout';
import { AbstractOperation } from 'mongodb/src/operations/operation';

/** @internal */
export class UpdateSearchIndexOperation extends AbstractOperation<void> {
  override SERVER_COMMAND_RESPONSE_TYPE = MongoDBResponse;
  private readonly collection: Collection;
  private readonly name: string;
  private readonly definition: Document;

  constructor(collection: Collection, name: string, definition: Document) {
    super();
    this.collection = collection;
    this.name = name;
    this.definition = definition;
    this.ns = collection.fullNamespace;
  }

  override get commandName() {
    return 'updateSearchIndex' as const;
  }

  override buildCommand(_connection: Connection, _session?: ClientSession): Document {
    const namespace = this.collection.fullNamespace;
    return {
      updateSearchIndex: namespace.collection,
      name: this.name,
      definition: this.definition
    };
  }

  override handleOk(_response: MongoDBResponse): void {
    // no response.
  }

  override buildOptions(timeoutContext: TimeoutContext): ServerCommandOptions {
    return { session: this.session, timeoutContext };
  }
}
