// mongodb/src/index.ts
import { Admin } from 'mongodb/src/admin';
import { OrderedBulkOperation } from 'mongodb/src/bulk/ordered';
import { UnorderedBulkOperation } from 'mongodb/src/bulk/unordered';
import { ChangeStream } from 'mongodb/src/change_stream';
import { Collection } from 'mongodb/src/collection';
import { AbstractCursor } from 'mongodb/src/cursor/abstract_cursor';
import { AggregationCursor } from 'mongodb/src/cursor/aggregation_cursor';
import { FindCursor } from 'mongodb/src/cursor/find_cursor';
import { ListCollectionsCursor } from 'mongodb/src/cursor/list_collections_cursor';
import { ListIndexesCursor } from 'mongodb/src/cursor/list_indexes_cursor';
import type { RunCommandCursor } from 'mongodb/src/cursor/run_command_cursor';
import { Db } from 'mongodb/src/db';
import { GridFSBucket } from 'mongodb/src/gridfs';
import { GridFSBucketReadStream } from 'mongodb/src/gridfs/download';
import { GridFSBucketWriteStream } from 'mongodb/src/gridfs/upload';
import { MongoClient } from 'mongodb/src/mongo_client';
import { CancellationToken } from 'mongodb/src/mongo_types';
import { ClientSession } from 'mongodb/src/sessions';

/** @public */
export { BSON } from 'mongodb/src/bson';
export {
  Binary,
  BSONRegExp,
  BSONSymbol,
  BSONType,
  Code,
  DBRef,
  Decimal128,
  Double,
  Int32,
  Long,
  MaxKey,
  MinKey,
  ObjectId,
  Timestamp,
  UUID
} from 'mongodb/src/bson';
export {
  type AnyBulkWriteOperation,
  type BulkWriteOptions,
  MongoBulkWriteError
} from 'mongodb/src/bulk/common';
export { ClientEncryption } from 'mongodb/src/client-side-encryption/client_encryption';
export { ChangeStreamCursor } from 'mongodb/src/cursor/change_stream_cursor';
export { ExplainableCursor } from 'mongodb/src/cursor/explainable_cursor';
export {
  MongoAPIError,
  MongoAWSError,
  MongoAzureError,
  MongoBatchReExecutionError,
  MongoChangeStreamError,
  MongoClientBulkWriteCursorError,
  MongoClientBulkWriteError,
  MongoClientBulkWriteExecutionError,
  MongoClientClosedError,
  MongoCompatibilityError,
  MongoCursorExhaustedError,
  MongoCursorInUseError,
  MongoDecompressionError,
  MongoDriverError,
  MongoError,
  MongoExpiredSessionError,
  MongoGCPError,
  MongoGridFSChunkError,
  MongoGridFSStreamError,
  MongoInvalidArgumentError,
  MongoKerberosError,
  MongoMissingCredentialsError,
  MongoMissingDependencyError,
  MongoNetworkError,
  MongoNetworkTimeoutError,
  MongoNotConnectedError,
  MongoOIDCError,
  MongoOperationTimeoutError,
  MongoParseError,
  MongoRuntimeError,
  MongoServerClosedError,
  MongoServerError,
  MongoServerSelectionError,
  MongoStalePrimaryError,
  MongoSystemError,
  MongoTailableCursorError,
  MongoTopologyClosedError,
  MongoTransactionError,
  MongoUnexpectedServerResponseError,
  MongoWriteConcernError,
  WriteConcernErrorResult
} from 'mongodb/src/error';
export {
  AbstractCursor,
  // Actual driver classes exported
  Admin,
  AggregationCursor,
  CancellationToken,
  ChangeStream,
  ClientSession,
  Collection,
  Db,
  FindCursor,
  GridFSBucket,
  GridFSBucketReadStream,
  GridFSBucketWriteStream,
  ListCollectionsCursor,
  ListIndexesCursor,
  MongoClient,
  OrderedBulkOperation,
  RunCommandCursor,
  UnorderedBulkOperation
};

// enums
export { BatchType } from 'mongodb/src/bulk/common';
export { AutoEncryptionLoggerLevel } from 'mongodb/src/client-side-encryption/auto_encrypter';
export { GSSAPICanonicalizationValue } from 'mongodb/src/cmap/auth/gssapi';
export { AuthMechanism } from 'mongodb/src/cmap/auth/providers';
export { Compressor } from 'mongodb/src/cmap/wire_protocol/compression';
export { CURSOR_FLAGS, CursorTimeoutMode } from 'mongodb/src/cursor/abstract_cursor';
export { MongoErrorLabel } from 'mongodb/src/error';
export { ExplainVerbosity } from 'mongodb/src/explain';
export { ServerApiVersion } from 'mongodb/src/mongo_client';
export { MongoLoggableComponent, SeverityLevel } from 'mongodb/src/mongo_logger';
export { ReturnDocument } from 'mongodb/src/operations/find_and_modify';
export { ProfilingLevel } from 'mongodb/src/operations/set_profiling_level';
export { ReadConcernLevel } from 'mongodb/src/read_concern';
export { ReadPreferenceMode } from 'mongodb/src/read_preference';
export { ServerType, TopologyType } from 'mongodb/src/sdam/common';

// Helper classes
export type { AWSCredentialProvider } from 'mongodb/src/cmap/auth/aws_temporary_credentials';
export type { AWSCredentials } from 'mongodb/src/deps';
export { ReadConcern } from 'mongodb/src/read_concern';
export { ReadPreference } from 'mongodb/src/read_preference';
export { WriteConcern } from 'mongodb/src/write_concern';
// events
export {
  CommandFailedEvent,
  CommandStartedEvent,
  CommandSucceededEvent
} from 'mongodb/src/cmap/command_monitoring_events';
export {
  ConnectionCheckedInEvent,
  ConnectionCheckedOutEvent,
  ConnectionCheckOutFailedEvent,
  ConnectionCheckOutStartedEvent,
  ConnectionClosedEvent,
  ConnectionCreatedEvent,
  ConnectionPoolClearedEvent,
  ConnectionPoolClosedEvent,
  ConnectionPoolCreatedEvent,
  ConnectionPoolMonitoringEvent,
  ConnectionPoolReadyEvent,
  ConnectionReadyEvent
} from 'mongodb/src/cmap/connection_pool_events';
export {
  ServerClosedEvent,
  ServerDescriptionChangedEvent,
  ServerHeartbeatFailedEvent,
  ServerHeartbeatStartedEvent,
  ServerHeartbeatSucceededEvent,
  ServerOpeningEvent,
  TopologyClosedEvent,
  TopologyDescriptionChangedEvent,
  TopologyOpeningEvent
} from 'mongodb/src/sdam/events';
export {
  ServerSelectionEvent,
  ServerSelectionFailedEvent,
  ServerSelectionStartedEvent,
  ServerSelectionSucceededEvent,
  WaitingForSuitableServerEvent
} from 'mongodb/src/sdam/server_selection_events';
export { SrvPollingEvent } from 'mongodb/src/sdam/srv_polling';

// type only exports below, these are removed from emitted JS
export type { AdminPrivate } from 'mongodb/src/admin';
export type { BSONElement, BSONSerializeOptions, Document } from 'mongodb/src/bson';
export type { deserialize, serialize } from 'mongodb/src/bson';
export type {
  BulkResult,
  BulkWriteOperationError,
  BulkWriteResult,
  DeleteManyModel,
  DeleteOneModel,
  InsertOneModel,
  ReplaceOneModel,
  UpdateManyModel,
  UpdateOneModel,
  WriteConcernError,
  WriteError
} from 'mongodb/src/bulk/common';
export type {
  Batch,
  BulkOperationBase,
  BulkOperationPrivate,
  FindOperators,
  WriteConcernErrorData
} from 'mongodb/src/bulk/common';
export type {
  ChangeStreamCollModDocument,
  ChangeStreamCreateDocument,
  ChangeStreamCreateIndexDocument,
  ChangeStreamDeleteDocument,
  ChangeStreamDocument,
  ChangeStreamDocumentCollectionUUID,
  ChangeStreamDocumentCommon,
  ChangeStreamDocumentKey,
  ChangeStreamDocumentOperationDescription,
  ChangeStreamDocumentWallTime,
  ChangeStreamDropDatabaseDocument,
  ChangeStreamDropDocument,
  ChangeStreamDropIndexDocument,
  ChangeStreamEvents,
  ChangeStreamInsertDocument,
  ChangeStreamInvalidateDocument,
  ChangeStreamNameSpace,
  ChangeStreamOptions,
  ChangeStreamRefineCollectionShardKeyDocument,
  ChangeStreamRenameDocument,
  ChangeStreamReplaceDocument,
  ChangeStreamReshardCollectionDocument,
  ChangeStreamShardCollectionDocument,
  ChangeStreamSplitEvent,
  ChangeStreamUpdateDocument,
  OperationTime,
  ResumeToken,
  UpdateDescription
} from 'mongodb/src/change_stream';
export type { AutoEncrypter } from 'mongodb/src/client-side-encryption/auto_encrypter';
export type { AutoEncryptionOptions } from 'mongodb/src/client-side-encryption/auto_encrypter';
export type { AutoEncryptionExtraOptions } from 'mongodb/src/client-side-encryption/auto_encrypter';
export type {
  AWSEncryptionKeyOptions,
  AzureEncryptionKeyOptions,
  ClientEncryptionCreateDataKeyProviderOptions,
  ClientEncryptionEncryptOptions,
  ClientEncryptionOptions,
  ClientEncryptionRewrapManyDataKeyProviderOptions,
  ClientEncryptionRewrapManyDataKeyResult,
  DataKey,
  GCPEncryptionKeyOptions,
  KMIPEncryptionKeyOptions,
  RangeOptions,
  TextQueryOptions
} from 'mongodb/src/client-side-encryption/client_encryption';
export {
  MongoCryptAzureKMSRequestError,
  MongoCryptCreateDataKeyError,
  MongoCryptCreateEncryptedCollectionError,
  MongoCryptError,
  MongoCryptInvalidArgumentError,
  MongoCryptKMSRequestNetworkTimeoutError
} from 'mongodb/src/client-side-encryption/errors';
export type { MongocryptdManager } from 'mongodb/src/client-side-encryption/mongocryptd_manager';
export type {
  AWSKMSProviderConfiguration,
  AzureKMSProviderConfiguration,
  ClientEncryptionDataKeyProvider,
  CredentialProviders,
  GCPKMSProviderConfiguration,
  KMIPKMSProviderConfiguration,
  KMSProviders,
  LocalKMSProviderConfiguration
} from 'mongodb/src/client-side-encryption/providers';
export type {
  ClientEncryptionSocketOptions,
  ClientEncryptionTlsOptions,
  CSFLEKMSTlsOptions,
  StateMachineExecutable
} from 'mongodb/src/client-side-encryption/state_machine';
export type { AuthContext, AuthProvider } from 'mongodb/src/cmap/auth/auth_provider';
export type {
  AuthMechanismProperties,
  MongoCredentials,
  MongoCredentialsOptions
} from 'mongodb/src/cmap/auth/mongo_credentials';
export type {
  IdPInfo,
  IdPServerResponse,
  OIDCCallbackFunction,
  OIDCCallbackParams,
  OIDCResponse
} from 'mongodb/src/cmap/auth/mongodb_oidc';
export type { Workflow } from 'mongodb/src/cmap/auth/mongodb_oidc';
export type { TokenCache } from 'mongodb/src/cmap/auth/mongodb_oidc/token_cache';
export type {
  MessageHeader,
  OpCompressedRequest,
  OpCompressesRequestOptions,
  OpMsgOptions,
  OpMsgRequest,
  OpMsgResponse,
  OpQueryOptions,
  OpQueryRequest,
  OpReply,
  WriteProtocolMessageType
} from 'mongodb/src/cmap/commands';
export type { HandshakeDocument } from 'mongodb/src/cmap/connect';
export type { LEGAL_TCP_SOCKET_OPTIONS, LEGAL_TLS_SOCKET_OPTIONS, Stream } from 'mongodb/src/cmap/connect';
export type {
  CommandOptions,
  Connection,
  ConnectionEvents,
  ConnectionOptions,
  ProxyOptions
} from 'mongodb/src/cmap/connection';
export type {
  ConnectionPool,
  ConnectionPoolEvents,
  ConnectionPoolOptions,
  PoolState,
  WaitQueueMember,
  WithConnectionCallback
} from 'mongodb/src/cmap/connection_pool';
export type { ClientMetadata } from 'mongodb/src/cmap/handshake/client_metadata';
export type { ConnectionPoolMetrics } from 'mongodb/src/cmap/metrics';
export type { StreamDescription, StreamDescriptionOptions } from 'mongodb/src/cmap/stream_description';
export type { CompressorName } from 'mongodb/src/cmap/wire_protocol/compression';
export type {
  JSTypeOf,
  OnDemandDocument,
  OnDemandDocumentDeserializeOptions
} from 'mongodb/src/cmap/wire_protocol/on_demand/document';
export type {
  CursorResponse,
  MongoDBResponse,
  MongoDBResponseConstructor
} from 'mongodb/src/cmap/wire_protocol/responses';
export type {
  CollectionOptions,
  CollectionPrivate,
  CountDocumentsOptions,
  ModifyResult
} from 'mongodb/src/collection';
export type {
  COMMAND_FAILED,
  COMMAND_STARTED,
  COMMAND_SUCCEEDED,
  CONNECTION_CHECK_OUT_FAILED,
  CONNECTION_CHECK_OUT_STARTED,
  CONNECTION_CHECKED_IN,
  CONNECTION_CHECKED_OUT,
  CONNECTION_CLOSED,
  CONNECTION_CREATED,
  CONNECTION_POOL_CLEARED,
  CONNECTION_POOL_CLOSED,
  CONNECTION_POOL_CREATED,
  CONNECTION_POOL_READY,
  CONNECTION_READY,
  MONGO_CLIENT_EVENTS,
  SERVER_CLOSED,
  SERVER_DESCRIPTION_CHANGED,
  SERVER_HEARTBEAT_FAILED,
  SERVER_HEARTBEAT_STARTED,
  SERVER_HEARTBEAT_SUCCEEDED,
  SERVER_OPENING,
  SERVER_SELECTION_FAILED,
  SERVER_SELECTION_STARTED,
  SERVER_SELECTION_SUCCEEDED,
  TOPOLOGY_CLOSED,
  TOPOLOGY_DESCRIPTION_CHANGED,
  TOPOLOGY_OPENING,
  WAITING_FOR_SUITABLE_SERVER
} from 'mongodb/src/constants';
export type {
  AbstractCursorEvents,
  AbstractCursorOptions,
  CursorFlag
} from 'mongodb/src/cursor/abstract_cursor';
export type {
  CursorTimeoutContext,
  InitialCursorResponse,
  InternalAbstractCursorOptions
} from 'mongodb/src/cursor/abstract_cursor';
export type { AggregationCursorOptions } from 'mongodb/src/cursor/aggregation_cursor';
export type { ChangeStreamCursorOptions } from 'mongodb/src/cursor/change_stream_cursor';
export type {
  ListSearchIndexesCursor,
  ListSearchIndexesOptions
} from 'mongodb/src/cursor/list_search_indexes_cursor';
export type { RunCursorCommandOptions } from 'mongodb/src/cursor/run_command_cursor';
export type { DbOptions, DbPrivate } from 'mongodb/src/db';
export type { Encrypter, EncrypterOptions } from 'mongodb/src/encrypter';
export type { AnyError, ErrorDescription, MongoNetworkErrorOptions } from 'mongodb/src/error';
export type {
  Explain,
  ExplainCommandOptions,
  ExplainOptions,
  ExplainVerbosityLike
} from 'mongodb/src/explain';
export type {
  GridFSBucketReadStreamOptions,
  GridFSBucketReadStreamOptionsWithRevision,
  GridFSBucketReadStreamPrivate,
  GridFSFile
} from 'mongodb/src/gridfs/download';
export type { GridFSBucketEvents, GridFSBucketOptions, GridFSBucketPrivate } from 'mongodb/src/gridfs';
export type { GridFSBucketWriteStreamOptions, GridFSChunk } from 'mongodb/src/gridfs/upload';
export type {
  Auth,
  DriverInfo,
  MongoClientEvents,
  MongoClientOptions,
  MongoClientPrivate,
  MongoOptions,
  PkFactory,
  ServerApi,
  SupportedNodeConnectionOptions,
  SupportedSocketOptions,
  SupportedTLSConnectionOptions,
  SupportedTLSSocketOptions,
  WithSessionCallback
} from 'mongodb/src/mongo_client';
export { MongoClientAuthProviders } from 'mongodb/src/mongo_client_auth_providers';
export type {
  Log,
  LogComponentSeveritiesClientOptions,
  LogConvertible,
  Loggable,
  LoggableCommandFailedEvent,
  LoggableCommandSucceededEvent,
  LoggableEvent,
  LoggableServerHeartbeatFailedEvent,
  LoggableServerHeartbeatStartedEvent,
  LoggableServerHeartbeatSucceededEvent,
  MongoDBLogWritable,
  MongoLogger,
  MongoLoggerEnvOptions,
  MongoLoggerMongoClientOptions,
  MongoLoggerOptions
} from 'mongodb/src/mongo_logger';
export type {
  Abortable,
  CommonEvents,
  EventsDescription,
  GenericListener,
  TypedEventEmitter
} from 'mongodb/src/mongo_types';
export type {
  AcceptedFields,
  AddToSetOperators,
  AlternativeType,
  ArrayElement,
  ArrayOperator,
  BitwiseFilter,
  BSONTypeAlias,
  Condition,
  EnhancedOmit,
  Filter,
  FilterOperations,
  FilterOperators,
  Flatten,
  InferIdType,
  IntegerType,
  IsAny,
  Join,
  KeysOfAType,
  KeysOfOtherType,
  MatchKeysAndValues,
  NestedPaths,
  NestedPathsOfType,
  NonObjectIdLikeDocument,
  NotAcceptedFields,
  NumericType,
  OneOrMore,
  OnlyFieldsOfType,
  OptionalId,
  OptionalUnlessRequiredId,
  PropertyType,
  PullAllOperator,
  PullOperator,
  PushOperator,
  RegExpOrString,
  RootFilterOperators,
  SchemaMember,
  SetFields,
  StrictFilter,
  StrictMatchKeysAndValues,
  StrictUpdateFilter,
  UpdateFilter,
  WithId,
  WithoutId
} from 'mongodb/src/mongo_types';
export type {
  AggregateOperation,
  AggregateOptions,
  DB_AGGREGATE_COLLECTION
} from 'mongodb/src/operations/aggregate';
export type {
  AnyClientBulkWriteModel,
  ClientBulkWriteError,
  ClientBulkWriteModel,
  ClientBulkWriteOptions,
  ClientBulkWriteResult,
  ClientDeleteManyModel,
  ClientDeleteOneModel,
  ClientDeleteResult,
  ClientInsertOneModel,
  ClientInsertOneResult,
  ClientReplaceOneModel,
  ClientUpdateManyModel,
  ClientUpdateOneModel,
  ClientUpdateResult,
  ClientWriteModel
} from 'mongodb/src/operations/client_bulk_write/common';
export type {
  CollationOptions,
  CommandOperation,
  CommandOperationOptions,
  OperationParent
} from 'mongodb/src/operations/command';
export type { CountOptions } from 'mongodb/src/operations/count';
export type {
  ClusteredCollectionOptions,
  CreateCollectionOptions,
  TimeSeriesCollectionOptions
} from 'mongodb/src/operations/create_collection';
export type { DeleteOptions, DeleteResult, DeleteStatement } from 'mongodb/src/operations/delete';
export type { DistinctOptions } from 'mongodb/src/operations/distinct';
export type { DropCollectionOptions, DropDatabaseOptions } from 'mongodb/src/operations/drop';
export type { EstimatedDocumentCountOptions } from 'mongodb/src/operations/estimated_document_count';
export type { FindOneOptions, FindOptions } from 'mongodb/src/operations/find';
export type {
  FindOneAndDeleteOptions,
  FindOneAndReplaceOptions,
  FindOneAndUpdateOptions
} from 'mongodb/src/operations/find_and_modify';
export type { IndexInformationOptions } from 'mongodb/src/operations/indexes';
export type {
  CreateIndexesOptions,
  DropIndexesOptions,
  IndexDescription,
  IndexDescriptionCompact,
  IndexDescriptionInfo,
  IndexDirection,
  IndexSpecification,
  ListIndexesOptions
} from 'mongodb/src/operations/indexes';
export type { InsertManyResult, InsertOneOptions, InsertOneResult } from 'mongodb/src/operations/insert';
export type { CollectionInfo, ListCollectionsOptions } from 'mongodb/src/operations/list_collections';
export type { ListDatabasesOptions, ListDatabasesResult } from 'mongodb/src/operations/list_databases';
export type { AbstractOperation, Hint, OperationOptions } from 'mongodb/src/operations/operation';
export type { ProfilingLevelOptions } from 'mongodb/src/operations/profiling_level';
export type { RemoveUserOptions } from 'mongodb/src/operations/remove_user';
export type { RenameOptions } from 'mongodb/src/operations/rename';
export type { RunCommandOptions } from 'mongodb/src/operations/run_command';
export type { SearchIndexDescription } from 'mongodb/src/operations/search_indexes/create';
export type { SetProfilingLevelOptions } from 'mongodb/src/operations/set_profiling_level';
export type { DbStatsOptions } from 'mongodb/src/operations/stats';
export type {
  ReplaceOptions,
  UpdateOptions,
  UpdateResult,
  UpdateStatement
} from 'mongodb/src/operations/update';
export type { ValidateCollectionOptions } from 'mongodb/src/operations/validate_collection';
export type { ReadConcernLike } from 'mongodb/src/read_concern';
export type {
  HedgeOptions,
  ReadPreferenceFromOptions,
  ReadPreferenceLike,
  ReadPreferenceLikeOptions,
  ReadPreferenceOptions
} from 'mongodb/src/read_preference';
export type { OsAdapter, Runtime, RuntimeAdapters } from 'mongodb/src/runtime_adapters';
export type { ClusterTime } from 'mongodb/src/sdam/common';
export type {
  Monitor,
  MonitorEvents,
  MonitorInterval,
  MonitorIntervalOptions,
  MonitorOptions,
  MonitorPrivate,
  RTTPinger,
  RTTPingerOptions,
  RTTSampler,
  ServerMonitoringMode
} from 'mongodb/src/sdam/monitor';
export type {
  Server,
  ServerCommandOptions,
  ServerEvents,
  ServerOptions,
  ServerPrivate
} from 'mongodb/src/sdam/server';
export type {
  ServerDescription,
  ServerDescriptionOptions,
  TagSet,
  TopologyVersion
} from 'mongodb/src/sdam/server_description';
export type { DeprioritizedServers, ServerSelector } from 'mongodb/src/sdam/server_selection';
export type { SrvPoller, SrvPollerEvents, SrvPollerOptions } from 'mongodb/src/sdam/srv_polling';
export type {
  ConnectOptions,
  SelectServerOptions,
  ServerSelectionCallback,
  ServerSelectionRequest,
  Topology,
  TopologyEvents,
  TopologyOptions,
  TopologyPrivate
} from 'mongodb/src/sdam/topology';
export type { TopologyDescription, TopologyDescriptionOptions } from 'mongodb/src/sdam/topology_description';
export type {
  ClientSessionEvents,
  ClientSessionOptions,
  EndSessionOptions,
  ServerSession,
  ServerSessionId,
  ServerSessionPool,
  WithTransactionCallback
} from 'mongodb/src/sessions';
export type { Sort, SortDirection, SortDirectionForCmd, SortForCmd } from 'mongodb/src/sort';
export type {
  CSOTTimeoutContext,
  CSOTTimeoutContextOptions,
  LegacyTimeoutContext,
  LegacyTimeoutContextOptions,
  Timeout,
  TimeoutContext,
  TimeoutContextOptions
} from 'mongodb/src/timeout';
export type { Transaction, TransactionOptions, TxnState } from 'mongodb/src/transactions';
export type {
  BufferPool,
  Callback,
  EventEmitterWithState,
  HostAddress,
  List,
  MongoDBCollectionNamespace,
  MongoDBNamespace
} from 'mongodb/src/utils';
export type { W, WriteConcernOptions, WriteConcernSettings } from 'mongodb/src/write_concern';
